modules/rgbd/src/hash_tsdf.cpp

```diff
@@ -317,19 +333,116 @@ inline Point3f HashTSDFVolumeCPU::getNormalVoxel(Point3f point) const
-
+#else
+        normal[c] = tsdfToFloat(at(Point3f(pointNext)).tsdf - at(Point3f(pointPrev)).tsdf);
+#endif
         pointPrev[c] = pointVec[c];
         pointNext[c] = pointVec[c];
     }
-    //std::cout << normal << std::endl;
+    
     float nv = sqrt(normal[0] * normal[0] +
                      normal[1] * normal[1] +
                      normal[2] * normal[2]);
     return nv < 0.0001f ? nan3 : normal/nv;
 }
 
+inline Point3f HashTSDFVolumeCPU::_getNormalVoxel(Point3f point) const
+{
+    cv::Point3f neighbourCoords[] = {
+                                Point3f(0, 0, 0),
+                                Point3f(0, 0, 1),
+                                Point3f(0, 1, 0),
+                                Point3f(0, 1, 1),
+                                Point3f(1, 0, 0),
+                                Point3f(1, 0, 1),
+                                Point3f(1, 1, 0),
+                                Point3f(1, 1, 1) };
+    for (int i = 0; i < 8; i++)
+        neighbourCoords[i] *= voxelSize;
+    
+    auto interpolate = [](cv::Point3f point, float vx[8])
+    {
+        int ix = cvFloor(point.x);
+        int iy = cvFloor(point.y);
+        int iz = cvFloor(point.z);
+
+        float tx = point.x - ix;
+        float ty = point.y - iy;
+        float tz = point.z - iz;
+
+        float v00 = vx[0] + tz * (vx[1] - vx[0]);
+        float v01 = vx[2] + tz * (vx[3] - vx[2]);
+        float v10 = vx[4] + tz * (vx[5] - vx[4]);
+        float v11 = vx[6] + tz * (vx[7] - vx[6]);
+
+        float v0 = v00 + ty * (v01 - v00);
+        float v1 = v10 + ty * (v11 - v10);
+
+        return v0 + tx * (v1 - v0);
+    };
+
+    Point3f pointVec(point);
+    Point3f normal = Point3f(0, 0, 0);
+
+    Point3f c0_pointPrev = pointVec, c0_pointNext = pointVec;
+    Point3f c1_pointPrev = pointVec, c1_pointNext = pointVec;
+    Point3f c2_pointPrev = pointVec, c2_pointNext = pointVec;
+    
+    c0_pointPrev.x -= voxelSize; c0_pointNext.x += voxelSize;
+    c1_pointPrev.y -= voxelSize; c1_pointNext.y += voxelSize;
+    c2_pointPrev.z -= voxelSize; c2_pointNext.z += voxelSize;
+
+    float c0_n_vx[8], c1_n_vx[8], c2_n_vx[8];
+    float c0_p_vx[8], c1_p_vx[8], c2_p_vx[8];
+
+    for (int i = 0; i < 8; i++)
+        c0_n_vx[i] = tsdfToFloat(at(neighbourCoords[i] + c0_pointNext).tsdf);
+    
+    for (int i = 0; i < 8; i++)
+        c0_p_vx[i] = tsdfToFloat(at(neighbourCoords[i] + c0_pointPrev).tsdf);
+    
+    for (int i = 0; i < 8; i++)
+        if (i % 4 > 1)
+            c1_n_vx[i] = tsdfToFloat(at(neighbourCoords[i] + c1_pointNext).tsdf);
+        else 
+            c1_p_vx[i] = tsdfToFloat(at(neighbourCoords[i] + c1_pointPrev).tsdf);
+
+    for (int i = 0; i < 8; i++)
+        if (i % 2 == 1)
+            c2_n_vx[i] = tsdfToFloat(at(neighbourCoords[i] + c2_pointNext).tsdf);
+        else
+            c2_p_vx[i] = tsdfToFloat(at(neighbourCoords[i] + c2_pointPrev).tsdf);
+
```

Comment:
The same about at inlining and hash table calls

Reply:
done